### ISSUE TITLE
Bug(qe): errors don't get logged

### ISSUE THREAD

**Issue by miguelff** [prisma/prisma#16226]
Supersedes https://github.com/prisma/prisma/issues/14933

### Background

Prisma can be configured to [log events to stdout](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/logging#log-to-stdout) or [emit logging events](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/logging#event-based-logging). Those events can be categorized into [four levels ](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#log-levels) which do not have a hierarchical relationship with each other. (A logging level does not include events from any other level)

One of these levels is `error`, but the [reference](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#log-levels) doesn't specify what is eligible to be logged using this level. 

Given that the reference has a [section on errors](https://www.prisma.io/docs/reference/api-reference/error-reference) users might have the reasonable expectation that, when any of those happen, an event is [logged to the console](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/logging#log-to-stdout) or [emitted](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/logging#event-based-logging) for subscription depending on the configuration. 

This is manifested in past issues like https://github.com/prisma/prisma/issues/14933, but given the reproduction scenarios below, the problem is not restricted to `findFirstOrThrow` and `findUniqueOrThrow` `NotFound` errors, but **to all the errors happening in Prisma**

### Reproduction

I created two reproduction scenarios, using the [libray engine](https://www.prisma.io/docs/reference/api-reference/environment-variables-reference#prisma_query_engine_binary) and the [binary engine](https://www.prisma.io/docs/reference/api-reference/environment-variables-reference#prisma_query_engine_binary) (As they exercise two different code paths for dispatching requests coming from the prisma client)

Both resulted on the error not being logged.

<details>
<summary>Premise: the engine was tricked to through a panic when dispatching a query</summary>

```diff
diff --git a/query-engine/core/src/interpreter/query_interpreters/read.rs b/query-engine/core/src/interpreter/query_interpreters/read.rs
index 6916f78fcc..878f876c5e 100644
--- a/query-engine/core/src/interpreter/query_interpreters/read.rs
+++ b/query-engine/core/src/interpreter/query_interpreters/read.rs
@@ -16,6 +16,10 @@ pub fn execute<'conn>(
     trace_id: Option<String>,
 ) -> BoxFuture<'conn, InterpretationResult<QueryResult>> {
     let fut = async move {
+        if 1 < 2 {
+            panic!("Damn that's a panic");
+        }
+
         match query {
             ReadQuery::RecordQuery(q) => read_one(tx, q, trace_id).await,
             ReadQuery::ManyRecordsQuery(q) => read_many(tx, q, trace_id).await,

```
</details>
<details>
<summary>Scenario 1 - Binary engine</summary>

`index.ts`
```typescript
import { PrismaClient, Prisma } from '.prisma/client'
​
async function main() {
  const prisma = new PrismaClient({
    log: [
      {
        emit: 'event',
        level: 'error',
      },    
    ],
    __internal: {
      engine: {
        endpoint: "http://127.0.0.1:57581",
      },
    },
  } as any);
​
  (prisma as PrismaClient<Prisma.PrismaClientOptions, 'error'>).$on("error", (e) => {
    console.log("WORKS WELL")
    process.exit(1)
  })
​
  const email = `user.${Date.now()}@prisma.io`
​
  const res = await prisma.user.findFirst({
    where: {
      email,
    }
  })
​
  console.log(res)
}
​
void main().catch((e) => {
  console.log("NOOOO!")
  console.log(e)
  process.exit(1)
})
```

`prisma.schema`
```schema
generator client {
  provider        = "prisma-client-js"
  output          = "../node_modules/.prisma/client"
  engineType      = "binary"
  previewFeatures = []
}
datasource db {
  provider = "postgresql"
  url      = "postgresql://postgres:prisma@localhost:5432"
}
model User {
  id    String @id @default(cuid())
  email String @unique
  posts Post[]
}
model Post {
  id       String @id @default(cuid())
  title    String
  author   User   @relation(fields: [authorId], references: [id])
  authorId String
}
```
Result:
<img width="957" alt="Screenshot 2022-11-10 at 16 19 09" src="https://user-images.githubusercontent.com/210307/201134274-55f9ead8-00d2-476e-9e8d-a344808aafaf.png">
</details>

<details>
<summary>Scenario 2 - Library engine</summary>

`index.ts`
```
import { PrismaClient, Prisma } from '.prisma/client'

async function main() {
  const prisma = new PrismaClient({
    log: [
      {
        emit: 'event',
        level: 'error',
      },
    ],
  });

  prisma.$on("error", (e) => {
    console.log("WORKS WELL")
    process.exit(1)
  })

  const email = `user.${Date.now()}@prisma.io`

  const res = await prisma.user.findFirst({
    where: {
      email,
    }
  })

  console.log(res)
}

void main().catch((e) => {
  console.log("NOOOO!")
  console.log(e)
  process.exit(1)
})
```

`prisma.schema`
```schema
generator client {
  provider        = "prisma-client-js"
  output          = "../node_modules/.prisma/client"
  previewFeatures = []
}
datasource db {
  provider = "postgresql"
  url      = "postgresql://postgres:prisma@localhost:5432"
}
model User {
  id    String @id @default(cuid())
  email String @unique
  posts Post[]
}
model Post {
  id       String @id @default(cuid())
  title    String
  author   User   @relation(fields: [authorId], references: [id])
  authorId String
}
```

Result
<img width="1074" alt="Screenshot 2022-11-10 at 15 41 33" src="https://user-images.githubusercontent.com/210307/201133883-0d4970de-c641-4e21-8e14-a766a484c7ee.png">
</details>

### Expected behavior

The specifics have to be determined. There's no previous record in the codebase to determine the correct behavior. And the [reference](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#log-levels)  of what constitutes an error is vague enough as to require a specification of how this should behave.

My proposal is to log with the error level what is said to constitute an error in the [error reference](https://www.prisma.io/docs/reference/api-reference/error-reference) thus following the principle of least astonishment. But I'd like to count with @janpio's take on how this should behave before jumping into implementation.

### Definition of done

- [x] Expected behavior clarified
- [x] Implementation and tests ensuring expected behavior
- [ ] Docs on the [log levels](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#log-levels) clarify the expected behavior

**Comment by ricardolpd** [comment]
We just upgraded prisma and found that the client logger was logging query errors. We found this a bit baffling as this should be handled by the service where the queries run from, and they should be logged or not depending on the business requirements. 

I would have expected that the client errors listener would log connection errors, internal errors from prisma, and other parts related to the query engine/pools and internals. Not query errors.

**Comment by janpio** [comment]
What would the "service where the queries run from" be in your interpretation? In my understanding there is only one component for Prisma Client users, which _is_ Prisma Client - where you can configure the logging in certain ways.

**Comment by ricardolpd** [comment]
@janpio we use insert as means to create idempontent requests, knowing that duplicate inserts will result in an error:

since the upgrade we see these `Unique constraint failed on the fields: (`transactionId`), target: transaction.create.`
being logged, however this is in a implementation choice from us, and it is an application error, not a prisma client error.

What i mean by this, is that service implementation caused the error, not the prisma client in some unrelated way from the application. 
The result of this design decision is that we are now filtering these messages through string comparisons (which is quite undesirable), when for me the application should choose if such errors are logged or not. Not the prisma client itself, as it removes control from the application, and generates quite a bit of undesired noise.

Hope my reply makes more sense.

**Comment by janpio** [comment]
Yeah I think I follow. What logging do you have enabled in Prisma Client?

**Comment by ricardolpd** [comment]
Probably easier to show it:

Query/Info are disabled in prod. I am not sure why the community wants this feature. But for us we would like to be able to exclude these errors.


```
prismaClient.$on('query' as any, (e: Prisma.QueryEvent) => {
  logger.trace(e, `[DB-Prisma] query: ${e.query}; duration: ${e.duration} ms.`);
});

prismaClient.$on('info' as any, (e: Prisma.LogEvent) => {
  logger.info(e, `[DB-Prisma] message: ${e.message}, target: ${e.target}.`);
});

prismaClient.$on('warn' as any, (e: Prisma.LogEvent) => {
  logger.warn(e, `[DB-Prisma] message: ${e.message}, target: ${e.target}.`);
});

const ignoredMessages = ['Unique constraint failed on the fields'];

const shouldLogEvent = (e: Prisma.LogEvent) => {
  if (
    ignoredMessages.some((ignoredMessage) => e.message.includes(ignoredMessage))
  ) {
    return false;
  }
  return true;
};

prismaClient.$on('error' as any, (e: Prisma.LogEvent) => {
  if (shouldLogEvent(e)) {
    logger.error(e, `[DB-Prisma] message: ${e.message}, target: ${e.target}.`);
  }
});

prismaClient.$on('beforeExit', async () => {
  await shutdownPromise;
  console.warn('[DB-Prisma] Prisma process is about to shutdown');
});
```

**Comment by janpio** [comment]
And via which of those are errors now unexpectedly logged?

**Comment by ricardolpd** [comment]
The error callback.

**Comment by janpio** [comment]
Ok, if you do not want the error caused by queries to be logged there, which would you then want to have logged? I am still trying to build a complete picture here.

**Comment by ricardolpd** [comment]
I will refer to my original [comment](https://github.com/prisma/prisma/issues/16226#issuecomment-1404826696):

> I would have expected that the client errors listener would log connection errors, internal errors from prisma, and other parts related to the query engine/pools and internals. Not query errors.

We didnt had this behavior in prisma in older versions. what caused the change to encompass this?

**Comment by janpio** [comment]
See from above:

> Given that the reference has a [section on errors](https://www.prisma.io/docs/reference/api-reference/error-reference) users might have the reasonable expectation that, when any of those happen, an event is [logged to the console](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/logging#log-to-stdout) or [emitted](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/logging#event-based-logging) for subscription depending on the configuration.
> 
> This is manifested in past issues like #14933, but given the reproduction scenarios below, the problem is not restricted to `findFirstOrThrow` and `findUniqueOrThrow` `NotFound` errors, but **to all the errors happening in Prisma**

**Comment by koutaro-masaki** [comment]
I am facing this same problem as @ricardolpd.

I updated my Prisma version from 3.15 to the latest and was baffled by the large amount of log output from tests that expect an exception. I tried to suppress it by passing `{ log: []} ` or `{ log: [ { 'level': 'error', 'emit': 'event' } ] }` to the constructor, but the log continued to output.

It looks very strange that the error log is output even when an exception is caught. Is there any way to prevent the error log from being output?
